### PR TITLE
Preliminary OCIO Code Generation Support  

### PR DESCRIPTION
**Pull request by bernardkwok**
Update #815.

### Basic Build Setup 
  - `MATERIALX_BUILD_OCIO` cmake option to enable OCIO building
  - `MATERIALX_OCIO_DIR` cmake option to incidate root directory of OCIO install
- New OCIO color management system with corresponding custom OCIO source code generator node logic
- New OCIO config option (_options.mtlx) on unit tests. Only GenGLSL and RenderGLSL unit tests take this into account.
- (To look at added an external project dependency if desired later)

### Basic Support
- Can generate code for LUT and non-LUT based transforms.
- LUT (texture based) transforms have no mechanism to prase and pass back image data for introspection. Delayed as config for base MTLX will most likely not have LUTs.
- Support for color3 inputs and outputs using color3->color4 conversion and color4->color3 conversion nodes connected to each OCIO shader node. This basically brackets the main function call.

### Test Results
  - Example generated GLSL code:
  ```
    vec4 c3_to_color3_acescg_value_cm_out = vec4(c3_to_color3_acescg_value_cm_in.x, c3_to_color3_acescg_value_cm_in.y, c3_to_color3_acescg_value_cm_in.z, 1);
    vec4 c3_to_color3_acescg2_value_cm_out = vec4(c3_to_color3_acescg2_value_cm_in.x, c3_to_color3_acescg2_value_cm_in.y, c3_to_color3_acescg2_value_cm_in.z, 1);
    vec4 color3_acescg_value_cm_out = IM_acescg_to_lin_rec709_color3_ocio(c3_to_color3_acescg_value_cm_out);
    vec4 color3_acescg2_value_cm_out = IM_acescg_to_lin_rec709_color3_ocio(c3_to_color3_acescg2_value_cm_out);
    vec3 color3_acescg_value_cm_to_c3_out = vec3(color3_acescg_value_cm_out.x, color3_acescg_value_cm_out.y, color3_acescg_value_cm_out.z);
    vec3 color3_acescg2_value_cm_to_c3_out = vec3(color3_acescg2_value_cm_out.x, color3_acescg2_value_cm_out.y, color3_acescg2_value_cm_out.z);
    vec3 color3_acescg_out = color3_acescg_value_cm_to_c3_out;
    vec3 color3_acescg2_out = color3_acescg2_value_cm_to_c3_out;
    vec3 add_acescg_out = color3_acescg_out + color3_acescg2_out;
  ```    
  where 2 transforms are bracketed and then routed to a add function to combine for final result.

   Single copy  of translform (in this case acescg to lin_rec709) looks like this (generated by OCIO code gen):
   ```
   // Declaration of the OCIO shader function
   vec4 IM_acescg_to_lin_rec709_color3_ocio(in vec4 inPixel)
   {
      vec4 outColor = inPixel;
  
       // Add a Matrix processing
  
      outColor = mat4(1.7049986809609248, -0.13025866763271018, -0.024007514741958086, 0., -0.62174765641381313, 1.1408020008115505, -0.12898266957287285, 0., -0.083251260888346112, -0.010544731974266384, 1.1529906045288774, 0., 0., 0., 0., 1.) * outColor;

     return outColor;
  } 
  ```
- Test renders (GLSL) for new unit test look as follows. 
image
  - Includes:
    - Test of color4 
    - Test of color3 where no conversion from color3->color4->color3 is required
     - Test reuse single update node with a given color transform (above add, and of transformed color3(0.5, 0,0, 0.0) and color3(0.0, 0.5, 0.0) -- resulting in yellow render)
     - Test use of multiple upstream nodes using the same color transform code but different inputs / outputs

### CODE REVIEW

File: source/MaterialXTest/MaterialXGenShader/GenShaderUtil.cpp [hunk 1]
```diff
@@ -449,6 +449,19 @@ void ShaderGeneratorTester::addColorManagement()
...
+
+#ifdef MATERIALX_BUILD_OCIO
+        mx::OCIOColorManagementSystemPtr ocio = mx::OCIOColorManagementSystem::create(target);
+        if (ocio)
+        {
+            ocio->setConfigFile("D:/Work/materialx/OpenColorIO-Config-ACES-1.2/aces_1.2/config.ocio");
+            ocio->loadLibrary(_dependLib);
+            mx::GenContext context(_shaderGenerator);
+            mx::ColorSpaceTransform transform("acescg", "lin_rec709", mx::Type::COLOR3);
+            ocio->createNode(nullptr, transform, "blah", context);
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXGenShader/GenShaderUtil.cpp:461]: Dummy call for testing.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    {
+        const ShaderGenerator& shadergen = context.getShaderGenerator();
+
+        // Declare the output variable
+        shadergen.emitLineBegin(stage);
+        shadergen.emitOutput(node.getOutput(0), true, false, context, stage);
+        shadergen.emitString(" = ", stage);
+
+        // Emit function name. TODO add vec3/vec4 conversion as needed
+        shadergen.emitString(_functionName + "(", stage);
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:305]: need to wrap this into a vec3->vec4(ocio_function)->vec3 call if this is being called to return a vec3 value.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+        // TODO : Create an implementation based on source code
+        // Note: A separate utility could take OCIO shader code and generate a set of MTLX implementations.
+        //       This could be useful to update the DefaultColorSystems implementations to maintain 
+        //       consistency with OCIO.
+        ImplementationPtr impl = _document->addImplementation(transformFunctionName);
+        if (impl)
+        {
+            // Q: How to parse inputs ?
+            impl->addInput("inPixel", transform.type->getName());
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:248]: need proper call to get the input.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    if (!_functionSource.empty())
+    {
+        const ShaderGenerator& shadergen = context.getShaderGenerator();
+        shadergen.emitBlock(_functionSource, context, stage);
+        shadergen.emitLineBreak(stage);
+    }
+    END_SHADER_STAGE(stage, Stage::PIXEL)
+}
+
+void OCIOSourceCodeNode::emitFunctionCall(const ShaderNode& node, GenContext& context, ShaderStage& stage) const
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:287]: todo: uniforms are not handled yet.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+ShaderNodeImplPtr OCIOSourceCodeNode::create()
+{
+    return std::make_shared<OCIOSourceCodeNode>();
+}
+
+void OCIOSourceCodeNode::emitFunctionDefinition(const ShaderNode&, GenContext& context, ShaderStage& stage) const
+{
+    // Write source code as is
+    BEGIN_SHADER_STAGE(stage, Stage::PIXEL)
+    if (!_functionSource.empty())
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:278]: todo: how to ensure secondary functions have unique names ?

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        nodeImpl = OCIOSourceCodeNode::create();
+        nodeImpl->initialize(*impl, context);
+        context.addNodeImplementation(impl->getName(), nodeImpl);
+    }
+
+    // Create the node.
+    ShaderNodePtr shaderNode = ShaderNode::create(parent, name, nodeImpl, ShaderNode::Classification::TEXTURE);
+
+    // Create ports on the node.
+    ShaderInput* input = shaderNode->addInput("in", transform.type);
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:148]: todo: check if input and output shader port names need to match the source code generation.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    _document = document;
+    if (!_document)
+    {
+        throw ExceptionShaderGenError("No library loaded for color management system");
+    }
+
+    TargetDefPtr targetDef = _document->getTargetDef(_target);
+    const StringVec targets = targetDef->getMatchingTargets();
+
+    for (const string& target : targets)
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:70]: Simplistic mapping from MTLX base target to OCIO language.

File: source/MaterialXGenShader/CMakeLists.txt [hunk 8]
```diff
@@ -1,13 +1,38 @@
...
+
+if(MATERIALX_BUILD_OCIO)
+    set(MATERIALX_OCIO_DIR ${MATERIALX_OCIO_DIR})
+    message(STATUS "OCIO root path: ${MATERIALX_OCIO_DIR}")
+
+    message(STATUS "OCIO include path: ${MATERIALX_OCIO_DIR}/include")
+    include_directories("${MATERIALX_OCIO_DIR}/include")
+ 
+    message(STATUS "OCIO library path: ${MATERIALX_OCIO_DIR}/lib")
+    file(GLOB OCIO_LIBRARIES "${MATERIALX_OCIO_DIR}/lib/*.lib") # ${CMAKE_STATIC_LIBRARY_SUFFIX}")
```
Comment (bernardkwok) [inline source/MaterialXGenShader/CMakeLists.txt:29]: For now use hard-coded paths vs a "find package" utility as none currently exists.
Todo: fix up library dependencies to use proper platform extensions.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+        {
+            OCIO::GetCurrentConfig();
+        }
+        if (_ocioInfo->config)
+        {
+            OCIO::ColorSpaceSetRcPtr colorSpaces = _ocioInfo->config->getColorSpaces(nullptr);
+            int colorSpaceCount = colorSpaces->getNumColorSpaces();
+            for (int i = 0; i < colorSpaceCount; i++)
+            {
+                _colorSpaceNames.insert(colorSpaces->getColorSpaceNameByIndex(i));
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:99]: Not currently used. Just for debugging. Could be used to find available transforms ?

File: source/MaterialXGenShader/OCIOColorManagementSystem.h [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+/// A shared pointer to a OCIOColorManagementSystem
+using OCIOColorManagementSystemPtr = shared_ptr<class OCIOColorManagementSystem>;
+
+/// @class OCIOColorManagementSystem
+/// Class for a color management system which uses OCIO.
+///
+class MX_GENSHADER_API OCIOColorManagementSystem : public ColorManagementSystem
+{
+  public:
+    virtual ~OCIOColorManagementSystem() { }
```
Comment (feldstj) [inline source/MaterialXGenShader/OCIOColorManagementSystem.h:30]: Don't you need to delete _ocioInfo?

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+bool OCIOColorManagementSystem::setConfigFile(const FilePath& configFile)
+{
+    if (_configFile != configFile && configFile.size())
+    {
+        _configFile = configFile;
+        _colorSpaceNames.clear();
+
+        _ocioInfo->config = OCIO::Config::CreateFromFile(_configFile.asString().c_str());
```
Comment (hodoulp) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:88]: It throws if there is a problem with the config file.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+        if (!_ocioInfo->config)
+        {
+            OCIO::GetCurrentConfig();
```
Comment (hodoulp) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:91]: We usually implement the config file discovery in that order:
1. OCIO::CreateFromEnv()
2. OCIO::CreateFromFile()
3. OCIO::CreateFromStream("<MatX custom config file>"). <- My suggestion
4. OCIO::CreateRaw()

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+
+/// Returns whether this color management system supports a provided transform
+bool OCIOColorManagementSystem::supportsTransform(const ColorSpaceTransform& transform) const
+{
+    if (_ocioInfo->language == OCIO_UNSUPPORT_LANGUAGE)
+    {
+        return false;
+    }
+
+    OCIO::ConstProcessorRcPtr processor = _ocioInfo->config->getProcessor(transform.sourceSpace.c_str(), transform.targetSpace.c_str());
```
Comment (hodoulp) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:115]: It throws if there is a problem.

**Review by hodoulp — CHANGES_REQUESTED** [review]
Great first step.
Now I'm waiting the next step :)

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp (hunk 12 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/OCIOColorManagementSystem.cpp:91]: @hodoulp : If it's okay, I'd like to defer this until later to decide this logic and do it at a higher level (also it's tbd what that logic should be). For now this is a utility class which only considers the source to be a  file. 
I can just fail for now if the config file is not found so there are no "hidden" side-effects.
BTW: I did not see any an API t read a config file to create a stream.

File: source/MaterialXGenShader/OCIOColorManagementSystem.h (hunk 10 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/OCIOColorManagementSystem.h:30]: Thanks for the catch.

File: resources/Materials/TestSuite/_options.mtlx [hunk 14]
```diff
@@ -111,6 +111,9 @@
 
     <!-- External testing: extra comma separated list of test root paths -->
     <input name="externalTestPaths" type="string" value="" />
+     
+    <!-- Color management configuration file. Only applicable if building with OCIO -->
+    <input name="colorManagementConfigFile" type="string" value="" />
```
Comment (bernardkwok) [inline resources/Materials/TestSuite/_options.mtlx:116]: An explicit config file is required unit tests to run. Otherwise the default CM system is used.

File: source/MaterialXGenShader/ColorManagementSystem.cpp [hunk 15]
```diff
@@ -88,4 +88,107 @@ ShaderNodePtr ColorManagementSystem::createNode(const ShaderGraph* parent, const
     return shaderNode;
 }
 
+
+void ColorManagementSystem::getPortConnections(ShaderGraph* graph, ShaderNode* colorTransformNode,
```
Comment (bernardkwok) [inline source/MaterialXGenShader/ColorManagementSystem.cpp:92]: @feldstj : new code to wrap up a transform now with pre / post type conversion.
Reused for both input and ouptut port connections by just returning the input and output shader port.

File: source/MaterialXGenShader/ColorManagementSystem.cpp [hunk 16]
```diff
@@ -88,4 +88,107 @@ ShaderNodePtr ColorManagementSystem::createNode(const ShaderGraph* parent, const
...
+        {
+            colorTransformNodeInput->makeConnection(convertToColor3Ptr->getOutput(0));
+            convertToColor4Ptr->getInput(0)->makeConnection(colorTransformNodeOutput);
+            inputToConnect = convertToColor3Ptr->getInput(0);
+            outputToConnect = convertToColor4Ptr->getOutput(0);
+        }
+    }
+}
+
+void ColorManagementSystem::connectNodeToShaderInput(ShaderGraph* graph, ShaderNodePtr colorTransformNodePtr, ShaderInput* shaderInput, GenContext& context)
```
Comment (bernardkwok) [inline source/MaterialXGenShader/ColorManagementSystem.cpp:141]: Move the input and output connection logic from shadergraph to the colorsystem.
Same logic as before, just uses a `getPortConnections` utility. If no conversion required then the logic is the same as what currently exists.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+        nodeImpl = OCIOSourceCodeNode::create();
+        nodeImpl->initialize(*impl, context);
+        context.addNodeImplementation(impl->getName(), nodeImpl);
+    }
+
+    // Create the node.
+    ShaderNodePtr shaderNode = ShaderNode::create(parent, name, nodeImpl, ShaderNode::Classification::TEXTURE);
+
+    // Create ports on the node.
+    ShaderInput* input = shaderNode->addInput("in", Type::COLOR4);
```
Comment (bernardkwok) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:178]: OCIO color transform node is always color4.

File: source/MaterialXGenShader/ColorManagementSystem.cpp [hunk 18]
```diff
@@ -71,11 +71,11 @@ ShaderNodePtr ColorManagementSystem::createNode(const ShaderGraph* parent, const
 
     // Create ports on the node.
     ShaderInput* input = shaderNode->addInput("in", transform.type);
-    if (transform.type == Type::COLOR3)
+    if (transform.type->getName() == Type::COLOR3->getName())
     {
         input->setValue(Value::createValue(Color3(0.0f, 0.0f, 0.0f)));
     }
-    else if (transform.type == Type::COLOR4)
+    else if (transform.type->getName() == Type::COLOR4->getName())
```
Comment (niklasharrysson) [inline source/MaterialXGenShader/ColorManagementSystem.cpp:78]: Curios why this change was done, here and above, as it's now comparing strings instead of just a pointer compare.

File: source/MaterialXGenShader/ColorManagementSystem.cpp (hunk 16 above)
Comment (niklasharrysson) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:141]: Nitpick: For consistency maybe use the same signature as above and make the parameter `ShaderNode* colorTransformNode` instead of a shared pointer (and remove the conversion from shared to raw pointer below).

Comment (niklasharrysson) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:141]: I like the cleanup of moving color management node setup from shadergraph class to colorsystem class. Maybe take that all the way and move `ShaderGraph::addColorTransformNode` as well, to keep everything contained to colorsystem classes?

File: source/MaterialXGenShader/ColorManagementSystem.h [hunk 19]
```diff
@@ -72,6 +78,23 @@ class MX_GENSHADER_API ColorManagementSystem
     /// Returns an implementation for a given transform
     virtual ImplementationPtr getImplementation(const ColorSpaceTransform& transform) const = 0;
 
+    /// For a given transform node and a target port type to connect to, determine the appropriate
+    /// input and output ports. If the transform node's port type does not match the target type
+    /// additional "conversion" nodes will created and the appropriate input and output ports on these 
+    /// nodes will be returned.
+    ///
+    /// Possible configirations include the first two cases where no additional conversion nodes are required, and
```
Comment (niklasharrysson) [inline source/MaterialXGenShader/ColorManagementSystem.h:86]: Typo: configirations

File: source/MaterialXGenShader/ColorManagementSystem.h [hunk 20]
```diff
@@ -72,6 +78,23 @@ class MX_GENSHADER_API ColorManagementSystem
...
+    /// the latter two cases where conversions are required
+    /// 
+    ///  color3 input  -> color3 transform node -> color3 output
+    ///  color4 input  -> color4 transform node -> color3 output
+    ///  color3 input -> color3-to-color4 conversion -> color4 transform node -> color4_to_color3 conversion -> color3 output
+    ///  color4 input -> color4_to_color3 conversion -> color3 transform node -> color3-to-color4 conversion -> color4 output
```
Comment (niklasharrysson) [inline source/MaterialXGenShader/ColorManagementSystem.h:92]: Out of curiosity, why do we need to allow having color transform nodes of unmatching type connected? Such auto-conversion handling is not supported elsewhere in MaterialX. But maybe there is a good reason for this?

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+    _ocioInfo->target = EMPTY_STRING;
+
+    _document = document;
+    if (!_document)
+    {
+        throw ExceptionShaderGenError("No library loaded for color management system");
+    }
+
+    TargetDefPtr targetDef = _document->getTargetDef(_target);
+    const StringVec targets = targetDef->getMatchingTargets();
```
Comment (niklasharrysson) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:82]: Could add a nullptr check on targetdDef, to fail with an exception instead of crashing if a targetdef for the target has not been loaded.

File: source/MaterialXGenShader/OCIOColorManagementSystem.h [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+
+  private:
+    OCIOInformation*  _ocioInfo;
+    FilePath          _configFile;
+    string            _target;
+};
+
+
+/// Extending the SourceCodeNode with requirements for OCIO.
+class MX_GENSHADER_API OCIOSourceCodeNode : public SourceCodeNode
```
Comment (niklasharrysson) [inline source/MaterialXGenShader/OCIOColorManagementSystem.h:79]: From the code above, in `OCIOColorManagementSystem::getImplementation`, it seems the source code for the node is retrieved dynamically in runtime from the OCIO system. If that is the case an alternative design here would be to use a ShaderNodeImpl with dynamic code generation instead of deriving from SourceCodeNode.

I think this would simplify things. At least one advantage is that you could handle the color3->color4->color3 conversions explicitly without having to insert extra conversion nodes.

**Comment by niklasharrysson** [comment]
Very cool to have OCIO support in the works!

File: source/MaterialXGenShader/ColorManagementSystem.cpp [hunk 23]
```diff
@@ -88,4 +88,107 @@ ShaderNodePtr ColorManagementSystem::createNode(const ShaderGraph* parent, const
     return shaderNode;
 }
 
+
+void ColorManagementSystem::getPortConnections(ShaderGraph* graph, ShaderNode* colorTransformNode,
+                                               const TypeDesc* targetType, GenContext& context,
+                                               ShaderInput*& inputToConnect, ShaderOutput*& outputToConnect)
+{
+    const std::string colorTransformNodeName = colorTransformNode->getName();
+    ShaderOutput* colorTransformNodeOutput = colorTransformNode->getOutput(0);
```
Comment (feldstj) [inline source/MaterialXGenShader/ColorManagementSystem.cpp:97]: Why not just do:
outputToConnect = colorTransformNode->getOutput(0);
inputToConnect = colorTransformNode->getInput(0);
?

File: source/MaterialXGenShader/ColorManagementSystem.cpp [hunk 24]
```diff
@@ -88,4 +88,107 @@ ShaderNodePtr ColorManagementSystem::createNode(const ShaderGraph* parent, const
...
+    {
+        throw ExceptionShaderGenError("Cannot connect color management node to shader input");
+    }
+
+    ShaderNode* colorTransformNode = colorTransformNodePtr.get();
+    ShaderInput* inputToConnect = nullptr;
+    ShaderOutput* outputToConnect = nullptr;
+
+    getPortConnections(graph, colorTransformNode, shaderInput->getType(), context, inputToConnect, outputToConnect);
+    if (inputToConnect && outputToConnect)
```
Comment (feldstj) [inline source/MaterialXGenShader/ColorManagementSystem.cpp:153]: I think you are missing braces here,

File: source/MaterialXGenShader/ColorManagementSystem.h (hunk 19 above)
Comment (feldstj) [reply source/MaterialXGenShader/ColorManagementSystem.h:86]: Also, should be: nodes will **be** created

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 25]
```diff
@@ -0,0 +1,331 @@
...
+
+    for (const string& target : targets)
+    {
+        if (_ocioInfo->languageMap.count(target))
+        {
+            // Note that we store the "base" target it instead of any derived target
```
Comment (feldstj) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:88]: Grammar

File: source/MaterialXGenShader/ColorManagementSystem.cpp (hunk 18 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:78]: I'm not sure why but this pointer compare seems to be "unstable" when running through Python tests as it get's missed. When a string compare it seems to be stable. Could be the value set is not exactly these pointers so it fails. The error is that "color3" is not valid in the exception thrown.

File: source/MaterialXGenShader/ColorManagementSystem.cpp (hunk 16 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:141]: I was thinking about this but thought it's a bit non explicit that adding a node + doing the connections.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp [hunk 26]
```diff
@@ -0,0 +1,331 @@
...
+        shadergen.emitBlock(_functionSource, context, stage);
+        shadergen.emitLineBreak(stage);
+    }
+    END_SHADER_STAGE(stage, Stage::PIXEL)
+}
+
+void OCIOSourceCodeNode::emitFunctionCall(const ShaderNode& node, GenContext& context, ShaderStage& stage) const
+{
+    // The main variation from the default "hardware source code node" logic is that OCIO always returns
+    // a value from function invocation, so modify caller logic to declare an local variable and assign it
```
Comment (feldstj) [inline source/MaterialXGenShader/OCIOColorManagementSystem.cpp:305]: Minor: **a** local variable not **an**

File: source/MaterialXGenShader/ColorManagementSystem.cpp (hunk 16 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:141]: I'll fix the consistency issue.

File: source/MaterialXGenShader/ColorManagementSystem.h (hunk 20 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.h:92]: OCIO is always color4 so for now I need to promote / demote as necessary.

File: source/MaterialXGenShader/OCIOColorManagementSystem.cpp (hunk 21 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/OCIOColorManagementSystem.cpp:82]: Yes. I'll fix this.

File: source/MaterialXGenShader/OCIOColorManagementSystem.h (hunk 22 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/OCIOColorManagementSystem.h:79]: I had originally tried put the conversion directly inside but it was much easier to just create the graphs as every shading language does the conversion differently. Also it is indeed just a single source string that is returned so is most like the source code node -- and hence the type conversions live outside. 

Perhaps moving to a general ShaderNodeImpl will come later on as it's under discussion with Patrick how to a have a more modular breakdown of the source code.

File: source/MaterialXGenShader/ColorManagementSystem.h (hunk 19 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.h:86]: Will fix.

File: source/MaterialXGenShader/ColorManagementSystem.cpp (hunk 23 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:97]: I used these two local variables later on when doing connections.

File: source/MaterialXGenShader/ColorManagementSystem.cpp (hunk 24 above)
Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:153]: Thanks for the catch. Will fix.

File: source/MaterialXGenShader/ColorManagementSystem.cpp (hunk 18 above)
Comment (niklasharrysson) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:78]: If that's the case it's a serious problem, as our codegen type system relies on these pointer comparison. It is what we use everywhere else for type checking, so it's surprising that it would fail in this case. So I think it's something we need to investigate further why it would fail.

Comment (bernardkwok) [reply source/MaterialXGenShader/ColorManagementSystem.cpp:78]: I have logged issue #1221 for this.

**Review by feldstj — APPROVED** [review]
